### PR TITLE
added error return for account.fetch() failing if not signed in #18

### PR DESCRIPTION
**Pull request by ianstalter123**
added error for account fetch in reference to below issue
(probably need to add a test for this)

https://github.com/hoodiehq/hoodie-client-account/issues/18#issuecomment-159693950

### CODE REVIEW

File: lib/fetch.js [hunk 1]
```diff
@@ -9,24 +10,29 @@ internals.fetchProperties = require('../utils/fetch-properties')
...
+  } else {
+    merge(state.session.account, properties)
+  }
 
-    internals.saveSession({
-      cacheKey: state.cacheKey,
-      session: state.session
-    })
-
-    return properties
```
Comment (gr2m) [inline lib/fetch.js]: can you keep the indentation as it was?

File: lib/fetch.js [hunk 2]
```diff
@@ -9,24 +10,29 @@ internals.fetchProperties = require('../utils/fetch-properties')
 internals.saveSession = require('../utils/save-session')
 
 function fetch (state, path) {
+  if (!state.session) {
+    var error = new Error('Not signed in')
+    error.name = 'UnauthenticatedError'
+    return Promise.reject(error)
+  }
```
Comment (gr2m) [inline lib/fetch.js]: :ok_hand:

**Comment by gr2m** [comment]
Looks good, thanks a bunch!

Would you mind adding a test?

**Comment by ianstalter123** [comment]
I was running npm test and it was giving indentation errors, let me check
into writing test.

2015-12-11 23:03 GMT-08:00 Gregor Martynus <notifications@github.com>:

**Comment by ianstalter123** [comment]
@gr2m     any thoughts on writing the test ? 
so basically session would be empty and it would return a promise?

**Comment by gr2m** [comment]
@ianstalter123 here is [the only test we currently have for the `fetch` method](https://github.com/hoodiehq/hoodie-client-account/blob/master/tests/unit/fetch-test.js#L7-L25):

```js
test('fetch', function (t) {
  simple.mock(internals, 'fetchProperties').resolveWith({})

  fetch({
    baseUrl: 'http://example.com',
    session: {
      id: 'abc4567'
    }
  }, 'path')

  t.deepEqual(internals.fetchProperties.lastCall.arg, {
    url: 'http://example.com/session/account',
    bearerToken: 'abc4567',
    path: 'path'
  }, 'calls fetchProperties with account url')

  simple.restore()
  t.end()
})
```

I’d do a very similar one, where we pass in state with only `baseUrl`, but without `session`. Something like this (I didn’t try, just out of my head)

```js
test('fetch without state', function (t) {
  fetch({
    baseUrl: 'http://example.com'
  }, 'path')

  .then(t.fail.bind(t, 'Must reject')
  .then(function (error) {
    t.is(error.name, 'UnauthenticatedError', 'rejects with UnauthenticatedError error')
  })
})
```

Does that help? Does it make sense to you?

**Comment by ianstalter123** [comment]
@gr2m yeah makes sense I'm just not familiar with syntax for your test framework here - node unit tests I guess

**Comment by gr2m** [comment]
We use [tape](https://www.npmjs.com/package/tape) for tests, it’s rather tiny. For mocking of methods we use [simple-mock](https://www.npmjs.com/package/simple-mock).

For unit tests, we expose internal methods like `fetchProperties` internally for mocking, like here: https://github.com/hoodiehq/hoodie-client-account/blob/master/lib/fetch.js#L7-L9

**Comment by ianstalter123** [comment]
@gr2m added the test here which passes but doesn't return error message for some reason

**Comment by gr2m** [comment]
ah I made a mistake in my code, the 2nd `.then` should be a `.catch`. And instead of the `t.end()` add a `t.plan(1)` because it’s an asynchronous test

This should work

```js
test('fetch without state', function (t) {
  t.plan(1)

  fetch({
    baseUrl: 'http://example.com'
  }, 'path')

  .then(t.fail.bind(t, 'Must reject'))
  .then(function (error) {
    t.equal(error.name, 'UnauthenticatedError', 'rejects with UnauthenticatedError error')
  })
})
```

Does it make sense?

**Comment by ianstalter123** [comment]
@gr2m , so t.plan(1) , is for the count of asynchronous calls returning  apparently
and for some reason you don't need a t.end as a result -- well it works updated it there in 
commit

**Comment by gr2m** [comment]
`t.plan(1)` means that you expect 1 assertion, and if the code is all run but no assertion passed or more than one, then it will raise an issue.

So is this good to merge now from your side? Looks good to me.

We have to reword the commit messages so that semantic-release understands them, but I can do that. I’d do two commits

1. commit the test files with `test: account.fetch() fails if signed out` message
1. commit the rest with `fix: account.fetch() fails if signed out` message

That will make `semantic-release` publish a new patch release on npm

**Comment by ianstalter123** [comment]
@gr2m ok - are you going to separate into 2 commits and merge or do I need to rebase or do something else.

**Comment by gr2m** [comment]
w00p, all good, merged via e577566 & 4cb4e8c

**Pull request closed without merging**